### PR TITLE
Fix tables.records generator when min_records >= 2 and table has no unique columns

### PR DESCRIPTION
**Pull request by JoshRosen**
This PR fixes a corner case bug in the `tables.records` generator: if `min_records >= 2` and the table schema has no unique columns then Hypothesis would fail to generate examples.

The problem is the `unique_by` function passed to `strategies.lists`: if a table has no unique columns then `to_unique_fields()` will return an empty tuple for every record, causing all records to have the same uniqueness key. Instead, I think we should pass `unique_by=None` when there are no unique columns.

Here, I added a regression test and implemented that fix.

(Thanks for the cool library! I'm experimenting with using Hypothesis for testing SQL database engines and your code has been useful for handling the data generation pieces.)

### CODE REVIEW

File: tests/test_tables/test_records.py [hunk 1]
```diff
@@ -20,6 +23,18 @@ def test_table_records_factory(table: Table) -> None:
...
+
+def test_table_records_lists_factory_for_table_without_primary_key(metadata: MetaData) -> None:
+    table = Table('numbers', metadata, Column('number', Integer))
+    # min_size = 2 is added as a regression test for a corner-case where the
+    # absence of a primary key prevented multiple records from being generated:
+    table_records_lists = records.lists_factory(table, min_size=2)
+    _do_test_table_records_lists_factory(table, table_records_lists)
+
+
+def _do_test_table_records_lists_factory(table: Table, table_records_lists: SearchStrategy) -> None:
```
Comment (JoshRosen) [inline tests/test_tables/test_records.py]: Here, I tried to factor out the common assertion logic for re-use across both the existing randomized test and my new test. I've tried to follow the style / conventions that I've seen used in this repo, but please feel free to edit / clean up if needed.

**Comment by lycantropos** [comment]
@JoshRosen: thanks for your contribution, let me take a look

File: hypothesis_sqlalchemy/records.py [hunk 2]
```diff
@@ -60,10 +60,15 @@ def lists_factory(columns: List[Column],
     def to_unique_fields(row: Tuple[Any, ...]) -> Tuple[Any, ...]:
         return tuple(row[index] for index in unique_indices)
 
+    if unique_indices:
+        unique_by = to_unique_fields
```
Comment (lycantropos) [inline hypothesis_sqlalchemy/records.py]: it is great to return to old code after awhile, now I see that we can replace `to_unique_fields` with [`operator.itemgetter`](https://docs.python.org/3/library/operator.html#operator.itemgetter) like
```python
from operator import itemgetter
...
if unique_indices:
    unique_by = itemgetter(*unique_indices)
else:
    unique_by = None
```

File: tests/test_tables/test_records.py [hunk 3]
```diff
@@ -20,6 +23,19 @@ def test_table_records_factory(table: Table) -> None:
 
 def test_table_records_lists_factory(table: Table) -> None:
     table_records_lists = records.lists_factory(table)
+    _do_test_table_records_lists_factory(table, table_records_lists)
+
+
+def test_table_records_lists_factory_for_table_without_primary_key() -> None:
+    metadata = MetaData()
+    table = Table('numbers', metadata, Column('number', Integer))
```
Comment (lycantropos) [inline tests/test_tables/test_records.py]: this can be a fixture in **tests.fixtures.tables** module
```python
@pytest.fixture(scope='function')
def table_without_unique_columns(metadata: MetaData) -> Table:
    return Table('numbers', metadata, Column('number', Integer))
```

File: tests/test_tables/test_records.py (hunk 1 above)
Comment (lycantropos) [reply tests/test_tables/test_records.py]: We can leave a single test (in the state as it was before) if we slightly improve our tests `tables` strategy (located at **tests.strategies.tables** module) to generate tables without unique columns as well like
```python
from operator import attrgetter

from hypothesis import strategies
from sqlalchemy import MetaData

from hypothesis_sqlalchemy import (columns,
                                   tables)

metadata = MetaData()
non_unique_columns = columns.non_primary_keys_factory(
        are_unique=strategies.just(False))
tables_without_unique_columns = tables.factory(
        metadatas=strategies.just(metadata),
        columns_lists=strategies.lists(non_unique_columns,
                                       unique_by=attrgetter('name')))
tables_with_unique_columns = tables.factory(
        metadatas=strategies.just(metadata))
tables = tables_without_unique_columns | tables_with_unique_columns
```

File: tests/test_tables/test_records.py (hunk 3 above)
Comment (lycantropos) [reply tests/test_tables/test_records.py]: if we generalize `tables` strategy as described [here](https://github.com/lycantropos/hypothesis_sqlalchemy/pull/9#discussion_r262797773), then there is no need in this fixture

File: tests/test_tables/test_records.py (hunk 1 above)
Comment (lycantropos) [reply tests/test_tables/test_records.py]: btw this is a common scenario I use in `Hypothesis`-based tests: if there is an edge-case related to generated data that I haven't thought of, then I should rewrite my strategies to handle it (and set `repeat` option in **setup.cfg** file to some big value like `1000` and see if some error arised locally and after that rewrite tested code and be sure the problem was handled)

**Comment by lycantropos** [comment]
I've changed target branch to `dev`, because there are some refactoring changes that I wanted to apply once something significant will be added, after that I'll [bump patch version](https://github.com/lycantropos/hypothesis_sqlalchemy#bumping-version) and upload new package to PyPI

File: tests/test_tables/test_records.py (hunk 1 above)
Comment (JoshRosen) [reply tests/test_tables/test_records.py]: This is a great suggestion.

I've been playing around with this and I've run into a surprising behavior with the proposed `tables_without_unique_columns` generator: it always seems to generate the same example when used with your `utils.example` function. For example:

```python
from tests.utils import example
print(example(tables_without_unique_columns))
print(example(tables_without_unique_columns))
```

prints

```
aaaaaaaa
aaaaaaaa
```

but

```
from tests.utils import example
print(example(tables_with_unique_columns))
print(example(tables_with_unique_columns))
```

gives me proper randomness like 

```
jfbwqyljylr
iiyebapcritaqtefhuep
```

I think this problem stems from how `example()` is defined:

```python
def example(strategy: SearchStrategy) -> Any:
    return find(specifier=strategy,
                condition=lambda x: True,
                settings=settings(phases=[Phase.generate],
                                  verbosity=Verbosity.quiet))
```

What I think happens is that `find()` tries to return a minimal example satisfying the condition (so we're trying to work around that by skipping `Phrase.shrink`), but in this new case it just so happens that every strategy's _first_ invocation successfully generates a value (i.e. the "zero" example tried by Conjecture just-so-happened to work), whereas this wasn't the case for tables with primary keys (some constraint must be excluding the zero example).

Upstream Hypothesis's `example()` has a nice solution for this in https://github.com/HypothesisWorks/hypothesis/commit/503dbd341aecbcffd2387db0d552daa930a649e6: simply ignore the first generated example. I'm going to try changing your `example()` to simply delegate to the Hypothesis method to see if that allows my regression test to fail.

Comment (lycantropos) [reply tests/test_tables/test_records.py]: yeah, that's true, I've ended up using this approach in newly created projects, forgot to update this one, we can use `find` function from [`topo` repo](https://github.com/lycantropos/topo/blob/master/tests/utils.py#L14)

File: hypothesis_sqlalchemy/records.py (hunk 2 above)
Comment (lycantropos) [reply hypothesis_sqlalchemy/records.py]: this is not quite true for a case with `len(unique_indices) == 1` (`itemgetter` will return just a single object, not a singleton `tuple`), so we can move function definition in the `if` body like
```python
if unique_indices:
    def unique_by(row: Tuple[Any, ...]) -> Tuple[Any, ...]:
        return tuple(row[index] for index in unique_indices)
else:
    unique_by = None
```

File: tests/test_tables/test_records.py (hunk 1 above)
Comment (JoshRosen) [reply tests/test_tables/test_records.py]: Update: even if I use `strategy.example()` everywhere, I'm still unable to reproduce the bug without making further test changes. I think this is because the `repeat` of `test_table_records_lists_factory` means that every invocation constructs a brand-new strategy instance (via `table_records_lists = records.lists_factory(table)`)  and then calls `.example()` on that, so we only exercise any given strategy exactly once and therefore miss out on testing more complex generated examples.

For example, the repetitions do end up ensuring that we test a variety of different table types but they never generate tables of `length > 1`.

I think that we can work around this problem by generating and testing multiple examples in each `test_table_records_lists_factory` call.

**Update 2:** the `repeat` turned out not to be the problem; see discussion below of why `min_size=2` is needed.

Comment (lycantropos) [reply tests/test_tables/test_records.py]: try to rewrite `tests.utils.example` function as
```python
from typing import Any

from hypothesis import (Phase,
                        core,
                        settings)
from hypothesis.errors import (NoSuchExample,
                               Unsatisfiable)
from hypothesis.searchstrategy import SearchStrategy


def example(strategy: SearchStrategy) -> Any:
    first_object_list = []

    def condition(object_: Any) -> bool:
        if first_object_list:
            return True
        else:
            first_object_list.append(object_)
            return False

    try:
        return core.find(strategy,
                         condition,
                         settings=settings(database=None,
                                           phases=tuple(set(Phase)
                                                        - {Phase.shrink})))
    except (NoSuchExample, Unsatisfiable) as search_error:
        try:
            result, = first_object_list
        except ValueError as unpacking_error:
            raise unpacking_error from search_error
        else:
            return result
```
and let me know if it works

File: tests/test_tables/test_records.py [hunk 4]
```diff
@@ -19,7 +19,7 @@ def test_table_records_factory(table: Table) -> None:
 
 
 def test_table_records_lists_factory(table: Table) -> None:
-    table_records_lists = records.lists_factory(table)
+    table_records_lists = records.lists_factory(table, min_size=2)
```
Comment (JoshRosen) [inline tests/test_tables/test_records.py]: It turns out that we need the `min_size=2` here, as part of the test case itself, in order to meaningfully regression test the bug fixed here: if I reintroduce the bug and omit this constraint then it'll keep generating tables of size 1 (since that's all it could generate).

File: tests/test_tables/test_records.py (hunk 1 above)
Comment (JoshRosen) [reply tests/test_tables/test_records.py]: Yep, that seems to work 👍 .

Out of curiosity, what's the motivation for using this instead of the built-in `example()` (whose [implementation](https://github.com/HypothesisWorks/hypothesis/blame/36f33b21b8ad2417e3c70e47c096414e17efca25/hypothesis-python/src/hypothesis/searchstrategy/strategies.py#L256) looks somewhat similar). Is there a corner-case in the built-in `example()` that your implementation avoids?

File: tests/test_tables/test_records.py (hunk 4 above)
Comment (lycantropos) [reply tests/test_tables/test_records.py]: that's strange, locally I have `table_records_lists` of various lengths without specifying `min_size`

Comment (JoshRosen) [reply tests/test_tables/test_records.py]: Did you try re-introducing the original bug? If I revert the actual behavior changes (`git checkout -f origin/master -- hypothesis_sqlalchemy/records.py`), keep the test changes, and set `min_size=1` then the updated tests **don't** fail despite the bug having been reintroduced.

Comment (lycantropos) [reply tests/test_tables/test_records.py]: E.g if I add statement
```python
    print(len(table_records),
          len(table.c), any(map(is_column_unique, table.c)))
```
and run tests I get something like
```
# generated records count | columns count | table has unique column
1 3 False
3 7 True
5 4 False
4 7 True
7 0 False
7 5 True
6 15 True
9 4 False
8 7 True
5 2 False
5 0 False
10 6 False
22 18 True
1 11 True
6 18 True
4 7 False
5 5 True
2 0 False
21 3 False
7 13 True
8 5 True
2 9 True
```

Comment (lycantropos) [reply tests/test_tables/test_records.py]: oh, I've finally got it: to reproduce error we really should specify `min_size`

Comment (lycantropos) [reply tests/test_tables/test_records.py]: I haven't mentioned it anywhere, but there is a codestyle that I follow (a draft of it can be found [here](https://gist.github.com/lycantropos/a3fdf8e98e5edb5c41ea0564b8337310)) and following it keyword arguments should be passed on a separate line each, like
```python
    table_records_lists = records.lists_factory(table, 
                                                min_size=2)
```
in our particular case (sorry for being scrupulous)

IMO this makes code more readable (*sparse is better than dense* afterall)

File: tests/test_tables/test_records.py (hunk 1 above)
Comment (lycantropos) [reply tests/test_tables/test_records.py]: Sorry for ignoring last question, answer is simple: mentioned [change](https://github.com/HypothesisWorks/hypothesis/commit/503dbd341aecbcffd2387db0d552daa930a649e6) wasn't there at earlier versions and I've wasted large amount of time figuring it out, so after that I've decided to write my own version and support it independently from `example` method
